### PR TITLE
Replace extensions rule with more general export rule. Extensions must be public.

### PR DESCRIPTION
**Pull request by ben-polinsky**
- Remove legacy `@preview` tag - it is meaningless. Was intended to convey beta status for extensions. Now, extensions must have a `@public` tag.
- Lint non-extension exported apis to ensure they include one and only one release tag. Lint rule renamed.
- Some refactoring and cleanup

Issue in core repo: https://github.com/iTwin/itwinjs-core/issues/6573

### CODE REVIEW

**Comment by ben-polinsky** [comment]
I see no reason why we shouldn't accept @alpha tags as well... we don't have a non-extension lint rule for enforcing a release tag. I assume we've left that up to the api-extractor, but we could catch errors before then if we enforce it in lint. 
(I suppose an @internal tag doesn't make any sense for extensions...)

In addition, the rule is confusing because it allows the end-user to configure certain tags, but then hardcodes the `beta` requirement (previously `preview`) when generating the extensions output. 

Recommendations:
- Remove hard coded requirement on @beta tag and export any apis which have the @extensions tag, perhaps with the exception of @internal
- Add rule to validate exported apis have a release tag.

Open to thoughts

**Comment by ben-polinsky** [comment]
> I see no reason why we shouldn't accept `@alpha` tags as well... we don't have a non-extension lint rule for enforcing a release tag. I assume we've left that up to the api-extractor, but we could catch errors before then if we enforce it in lint. (I suppose an `@internal` tag doesn't make any sense for extensions...)
> 
> In addition, the rule is confusing because it allows the end-user to configure certain tags, but then hardcodes the `@beta` requirement (previously `preview`) when generating the extensions output.
> 
> Recommendations:
> 
> * Remove hard coded requirement on `@beta` tag and export any apis which have the `@extensions` tag, perhaps with the exception of `@internal`
> * Add rule to validate exported apis have a release tag.
> 
> Open to thoughts

Both of the above have been added in my latest changes. I have a cospace with core I am validating before merge.

**Comment by ben-polinsky** [comment]
@anmolshres98 @aruniverse we're good to go here, if you want to review.

File: dist/rules/exports-require-release-tags.js
Comment (anmolshres98) [inline dist/rules/exports-require-release-tags.js:17]: why is ClassDeclaration friendly name `real`? and why is that the default. I'm not sure I understand

File: dist/rules/exports-require-release-tags.js
Comment (anmolshres98) [inline dist/rules/exports-require-release-tags.js:43]: I think this needs to be worded a little better, even though that seems to have been the wording before

File: dist/rules/exports-require-release-tags.js
Comment (anmolshres98) [reply dist/rules/exports-require-release-tags.js:17]: I guess on second thought, you didn't originally write this, but do you happen to know? genuinely curious

File: tests/exports-require-release-tags.js
Comment (anmolshres98) [inline tests/exports-require-release-tags.js:123]: not sure I understand what this means. isn't `"destroyAllIModels"` a function and not a type?

**Review by anmolshres98 — COMMENTED** [review]
looks good except a couple clarification comments

File: dist/rules/exports-require-release-tags.js
Comment (ben-polinsky) [reply dist/rules/exports-require-release-tags.js:17]: I was also mystified by this. My best guess is `class` caused an issue somewhere as a reserved word and they went with... real? Commit history obscured as repo was moved from core. Looking into it as I am very curious.

Comment (ben-polinsky) [reply dist/rules/exports-require-release-tags.js:17]: No comment on the PR... @aruniverse remember anything about this?

File: dist/rules/exports-require-release-tags.js
Comment (ben-polinsky) [reply dist/rules/exports-require-release-tags.js:43]: Which one specifically?

File: tests/exports-require-release-tags.js
Comment (ben-polinsky) [reply tests/exports-require-release-tags.js:123]: Everything in TS has a type, so I went with that, but maybe it's enough to just say `destroyAllIModels`

File: dist/rules/exports-require-release-tags.js
Comment (anmolshres98) [reply dist/rules/exports-require-release-tags.js:43]: the `namespace`

File: tests/exports-require-release-tags.js
Comment (anmolshres98) [reply tests/exports-require-release-tags.js:123]: Did you mean `Check type of "destroyAllIModels"`? because `destroyAllIModels` isn't name of a type in core repo or anything like that right? correct me if I am wrong

File: dist/rules/exports-require-release-tags.js
Comment (anmolshres98) [reply dist/rules/exports-require-release-tags.js:17]: ya, weird. you can resolve this thread at your discretion

File: tests/exports-require-release-tags.js
Comment (ben-polinsky) [reply tests/exports-require-release-tags.js:123]: Right no, it's just a made up function for the test case. 
I did mean `check type of {{}}` I think that makes more sense.

File: dist/rules/exports-require-release-tags.js
Comment (ben-polinsky) [reply dist/rules/exports-require-release-tags.js:43]: Indeed will come up with something

File: tests/exports-require-release-tags.js
Comment (ben-polinsky) [reply tests/exports-require-release-tags.js]: Ah, I had hard coded `destroyAllIModels`... anyway, reworded

File: dist/rules/exports-require-release-tags.js
Comment (ben-polinsky) [reply dist/rules/exports-require-release-tags.js]: give it a looksee

Comment (anmolshres98) [reply dist/rules/exports-require-release-tags.js]: much better now, thanks

File: tests/exports-require-release-tags.js
Comment (anmolshres98) [reply tests/exports-require-release-tags.js]: I think I would prefer something like `Please review the tags for "destroyAllIModels" for reference.` or `Please review the tags for "destroyAllIModels" for example.` but the way it is I am not opposed to it either. Thanks for changing

**Review by anmolshres98 — APPROVED** [review]

**Review by aruniverse — APPROVED** [review]
lgtm, just add a beachball change so you can publish a new pre-release

**Comment by ben-polinsky** [comment]
> lgtm, just add a beachball change so you can publish a new pre-release

Have a change. I did set it to major, figured we could finally release 4.0.0.

**Comment by aruniverse** [comment]
> > lgtm, just add a beachball change so you can publish a new pre-release
> 
> Have a change. I did set it to major, figured we could finally release 4.0.0.

my bad; missed that. sounds good.

**Comment by ben-polinsky** [comment]
> > > > lgtm, just add a beachball change so you can publish a new pre-release
> > > 
> > > 
> > > Have a change. I did set it to major, figured we could finally release 4.0.0.
> > 
> > 
> > my bad; missed that. sounds good.
> 
> It didn't even take a year 🎉 (but hurry up, you only have a month left 😁)
> 
> Jokes aside, perhaps it would be a good time to remove style rules and switch to prettier? On the other hand, I wouldn't really mind another major bump in a month or so.

I agree - let's get crazy and release another major soon.

**Comment by paulius-valiunas** [comment]
Well, I don't want to hold up this release because it's long overdue... but then I also don't want to wait another year just because for some reason we want to keep an arbitrary interval between major versions. These are lint rules, it's not like we have to frequently patch vulnerabilities in previous major versions. We can release them as often as we want and let people update at their own pace.

**Pull request merged**
